### ISSUE TITLE
crash on reference handling on Psalm 5

### ISSUE THREAD

**Issue by orklah** [vimeo/psalm#7586]
I got a crash while running Psalm 5 on my work project

Here's the trace:
```
Uncaught TypeError: Argument 2 passed to Psalm\Internal\Provider\NodeDataProvider::setType() must be an instance of Psalm\Type\Union, instance of stdClass given, called in C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Fetch\VariableFetchAnalyzer.php on line 366 and defined in C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Provider\NodeDataProvider.php:53

Stack trace:
#0 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Fetch\VariableFetchAnalyzer.php(366): Psalm\Internal\Provider\NodeDataProvider->setType(Object(PhpParser\Node\Expr\Variable), Object(stdClass))

#1 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(165): Psalm\Internal\Analyzer\Statements\Expression\Fetch\VariableFetchAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Variable), Object(Psalm\Context), false, NULL, false, false, false)

#2 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(85): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::handleExpression(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Variable), Object(Psalm\Context), false, NULL, false, NULL, false)

#3 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Fetch\ArrayFetchAnalyzer.php(157): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Variable), Object(Psalm\Context))

#4 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(343): Psalm\Internal\Analyzer\Statements\Expression\Fetch\ArrayFetchAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\ArrayDimFetch), Object(Psalm\Context))

#5 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(85): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::handleExpression(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\ArrayDimFetch), Object(Psalm\Context), false, NULL, false, NULL, false)

#6 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\CastAnalyzer.php(62): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\ArrayDimFetch), Object(Psalm\Context))

#7 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(350): Psalm\Internal\Analyzer\Statements\Expression\CastAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Cast\Int_), Object(Psalm\Context))

#8 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(85): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::handleExpression(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Cast\Int_), Object(Psalm\Context), false, NULL, false, NULL, false)

#9 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Call\ArgumentsAnalyzer.php(242): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Cast\Int_), Object(Psalm\Context), false, NULL, false, NULL)

#10 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\CallAnalyzer.php(343): Psalm\Internal\Analyzer\Statements\Expression\Call\ArgumentsAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Array, Array, 'cm_attribute::g...', true, Object(Psalm\Context), Object(Psalm\Internal\Type\TemplateResult))

#11 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Call\StaticMethod\ExistingAtomicStaticCallAnalyzer.php(199): Psalm\Internal\Analyzer\Statements\Expression\CallAnalyzer::checkMethodArgs(Object(Psalm\Internal\MethodIdentifier), Array, Object(Psalm\Internal\Type\TemplateResult), Object(Psalm\Context), Object(Psalm\CodeLocation), Object(Psalm\Internal\Analyzer\StatementsAnalyzer))

#12 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Call\StaticMethod\AtomicStaticCallAnalyzer.php(855): Psalm\Internal\Analyzer\Statements\Expression\Call\StaticMethod\ExistingAtomicStaticCallAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\StaticCall), Object(PhpParser\Node\Identifier), Array, Object(Psalm\Context), Object(Psalm\Type\Atomic\TNamedObject), Object(Psalm\Internal\MethodIdentifier), 'cm_attribute::g...', Object(Psalm\Storage\ClassLikeStorage), false, NULL)

#13 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Call\StaticMethod\AtomicStaticCallAnalyzer.php(206): Psalm\Internal\Analyzer\Statements\Expression\Call\StaticMethod\AtomicStaticCallAnalyzer::handleNamedCall(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\StaticCall), Object(PhpParser\Node\Identifier), Object(Psalm\Context), Object(Psalm\Type\Atomic\TNamedObject), Array, 'cm_attribute', false, true, NULL)

#14 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Call\StaticCallAnalyzer.php(216): Psalm\Internal\Analyzer\Statements\Expression\Call\StaticMethod\AtomicStaticCallAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\StaticCall), Object(Psalm\Context), Object(Psalm\Type\Atomic\TNamedObject), false, false, false, true, NULL)

#15 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(208): Psalm\Internal\Analyzer\Statements\Expression\Call\StaticCallAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\StaticCall), Object(Psalm\Context), NULL)

#16 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(85): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::handleExpression(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\StaticCall), Object(Psalm\Context), false, NULL, false, NULL, false)

#17 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\AssignmentAnalyzer.php(245): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\StaticCall), Object(Psalm\Context))

#18 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(185): Psalm\Internal\Analyzer\Statements\Expression\AssignmentAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Variable), Object(PhpParser\Node\Expr\StaticCall), NULL, Object(Psalm\Context), NULL, Array, NULL)

#19 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(85): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::handleExpression(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Assign), Object(Psalm\Context), false, NULL, true, NULL, false)

#20 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(573): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Assign), Object(Psalm\Context), false, NULL, true)

#21 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(207): Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\Expression), Object(Psalm\Context), NULL)

#22 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\LoopAnalyzer.php(204): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context))

#23 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\ForeachAnalyzer.php(365): Psalm\Internal\Analyzer\Statements\Block\LoopAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Array, Array, Array, Object(Psalm\Internal\Scope\LoopScope), Object(Psalm\Context), false, false)

#24 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(531): Psalm\Internal\Analyzer\Statements\Block\ForeachAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\Foreach_), Object(Psalm\Context))

#25 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(207): Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\Foreach_), Object(Psalm\Context), NULL)

#26 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfElse\IfAnalyzer.php(170): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context))

#27 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfElseAnalyzer.php(267): Psalm\Internal\Analyzer\Statements\Block\IfElse\IfAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Internal\Scope\IfScope), Object(Psalm\Internal\Scope\IfConditionalScope), Object(Psalm\Context), Object(Psalm\Context), Array)

#28 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(519): Psalm\Internal\Analyzer\Statements\Block\IfElseAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context))

#29 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(207): Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context), NULL)

#30 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfElse\IfAnalyzer.php(170): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context))

#31 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfElseAnalyzer.php(267): Psalm\Internal\Analyzer\Statements\Block\IfElse\IfAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Internal\Scope\IfScope), Object(Psalm\Internal\Scope\IfConditionalScope), Object(Psalm\Context), Object(Psalm\Context), Array)

#32 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(519): Psalm\Internal\Analyzer\Statements\Block\IfElseAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context))

#33 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(207): Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context), NULL)

#34 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfElse\IfAnalyzer.php(170): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context))

#35 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfElseAnalyzer.php(267): Psalm\Internal\Analyzer\Statements\Block\IfElse\IfAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Internal\Scope\IfScope), Object(Psalm\Internal\Scope\IfConditionalScope), Object(Psalm\Context), Object(Psalm\Context), Array)

#36 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(519): Psalm\Internal\Analyzer\Statements\Block\IfElseAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context))

#37 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(207): Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context), NULL)

#38 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\LoopAnalyzer.php(204): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context))

#39 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\ForAnalyzer.php(91): Psalm\Internal\Analyzer\Statements\Block\LoopAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Array, Array, Array, Object(Psalm\Internal\Scope\LoopScope), Object(Psalm\Context))

#40 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(527): Psalm\Internal\Analyzer\Statements\Block\ForAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\For_), Object(Psalm\Context))

#41 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(207): Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\For_), Object(Psalm\Context), NULL)

#42 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfElse\IfAnalyzer.php(170): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context))

#43 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfElseAnalyzer.php(267): Psalm\Internal\Analyzer\Statements\Block\IfElse\IfAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Internal\Scope\IfScope), Object(Psalm\Internal\Scope\IfConditionalScope), Object(Psalm\Context), Object(Psalm\Context), Array)

#44 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(519): Psalm\Internal\Analyzer\Statements\Block\IfElseAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context))

#45 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(207): Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context), Object(Psalm\Context))

#46 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\FunctionLikeAnalyzer.php(475): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context), Object(Psalm\Context), true)

#47 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\ClassAnalyzer.php(1781): Psalm\Internal\Analyzer\FunctionLikeAnalyzer->analyze(Object(Psalm\Context), Object(Psalm\Internal\Provider\NodeDataProvider), Object(Psalm\Context))

#48 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\ClassAnalyzer.php(422): Psalm\Internal\Analyzer\ClassAnalyzer->analyzeClassMethod(Object(PhpParser\Node\Stmt\ClassMethod), Object(Psalm\Storage\ClassLikeStorage), Object(Psalm\Internal\Analyzer\ClassAnalyzer), Object(Psalm\Context), Object(Psalm\Context))

#49 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\FileAnalyzer.php(229): Psalm\Internal\Analyzer\ClassAnalyzer->analyze(Object(Psalm\Context), Object(Psalm\Context))

#50 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Codebase\Analyzer.php(360): Psalm\Internal\Analyzer\FileAnalyzer->analyze()

#51 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Codebase\Analyzer.php(617): Psalm\Internal\Codebase\Analyzer->Psalm\Internal\Codebase\{closure}(8291, 'C:\\Path...')

#52 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Codebase\Analyzer.php(289): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 1)

#53 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\ProjectAnalyzer.php(686): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 1, false, true)

#54 C:\Path\vendor\vimeo\psalm\src\Psalm\Internal\Cli\Psalm.php(376): Psalm\Internal\Analyzer\ProjectAnalyzer->check('C:\\Path...', true)

#55 C:\Path\vendor\vimeo\psalm\psalm(9): Psalm\Internal\Cli\Psalm::run(Array)

#56 {main}

(Psalm dev-master@cfce264bbaf1eedeacc752813ff383d8323765ce crashed due to an uncaught Throwable)
```

Running --debug-by-line on it reveal that the lines it crashes on look like

```php
$a_atti = &$this->list_serialise_in[$i][$ai];
$do_att = cm_attribute::get_from_id((int)$a_atti['key']);
```

It fails on the second line.

I'm pretty sure this has to do with references. I thought I took all I needed to have a small reproducer but I miss parent classes so I can't provide one for now

@AndrolGenhald The error seems pretty distinctive, I don't get how we could come to a point where a stdClass manages to enter the type system like that. Does it ring a bell? Maybe a polluted context?

**Comment by AndrolGenhald** [comment]
That's a real odd one, I have no idea how a `stdClass` would be getting in there. Did you run this before or after #7580 was fixed? I'd guess it's either a duplicate of that, or it's caused by the fix for it.

**Comment by orklah** [comment]
probably before. I could try to run it again

**Comment by AndrolGenhald** [comment]
Could you add in a print statement at VariableFetchAnalyzer.php:365 to see what `$var_name` is? I'd love to know how a `stdClass` is getting into `$vars_in_scope` in the first place. I'd love to debug this but I'm not having any luck reproducing so far.

**Comment by orklah** [comment]
It seems better after updating. I'll let the analysis finish but I'm pretty sure it passed the point of crash (the analysis takes 11 minutes when not in debug mode, and it's longer when it prints out all the output...)

**Comment by AndrolGenhald** [comment]
It still bothers me a bit that a `stdClass` got in there somehow, I would have expected it to be the same as the other one where both the reference and the referenced variable have their `$vars_in_scope` set to `null`, but if that fixed it I think it's probably safe now.

**Comment by orklah** [comment]
Oh nice!

I can confirm it's fixed :)

Also I was able to run our freak code through Psalm 5 without any other crash, that's a good sign :D
